// flex-embed/index.js
import React, {Component} from 'react';
import {Block} from 'jsxstyle';
import PropTypes from 'prop-types';

/**
 * Responsive, intrinsic ratio or percentage embeds
 */
export default class FlexEmbed extends Component {
  static propTypes = {
    /**
     * Ratio for embed innards, e.g. `3 / 4`, `9 / 16`, `0.75`
     */
    ratio: PropTypes.number,
    /**
     * @ignore
     */
    setRef: PropTypes.func,
    /**
     * @ignore
     */
    children: PropTypes.any,
    /**
     * Percentage of embed innards' width / height ratio, e.g. 75%
     */
    percentage: PropTypes.string
  };

  static defaultProps = {
    ratio: 1
  };

  render() {
    const {ratio, setRef, children, percentage, ...props} = this.props;

    return (
      <Block overflow="hidden" position="relative" {...props}>
        <Block width="100%" paddingBottom={percentage || `${ratio * 100}%`} />

        <Block
          top="0"
          left="0"
          props={{ref: setRef}}
          width="100%"
          height="100%"
          bottom="0"
          position="absolute">
          {children}
        </Block>
      </Block>
    );
  }
}
